test-app/tests/integration/components/toucan-form/toucan-form-test.gts

```diff
@@ -42,37 +42,75 @@ module('Integration | Component | ToucanForm', function (hooks) {
   });
 
   test('it sets the yielded component values based on `@data`', async function (assert) {
-    const data: { comment?: string } = { comment: 'textarea' };
+    const data: { comment?: string; firstName?: string } = {
+      comment: 'textarea',
+      firstName: 'input',
+    };
 
     await render(<template>
       <ToucanForm @data={{data}} as |form|>
         <form.Textarea @label="Comment" @name="comment" data-textarea />
+        <form.Input @label="Input" @name="firstName" data-input />
       </ToucanForm>
     </template>);
 
     assert.dom('[data-textarea]').hasAttribute('name', 'comment');
     assert.dom('[data-textarea]').hasValue('textarea');
+
+    assert.dom('[data-input]').hasAttribute('name', 'firstName');
+    assert.dom('[data-input]').hasValue('input');
   });
 
   test('it triggers validation and shows error messages in the Toucan Core components', async function (assert) {
-    const data: { comment?: string } = {};
-
-    const formValidateCallback = ({ comment }: { comment?: string }) =>
-      !comment
-        ? {
-            comment: [
-              {
-                type: 'uppercase',
-                value: comment,
-                message: 'Comment is required',
-              },
-            ],
-          }
-        : undefined;
+    const data: { comment?: string; firstName?: string } = {};
+
+    const formValidateCallback = ({
+      comment,
+      firstName,
+    }: {
+      comment?: string;
+      firstName?: string;
+    }) => {
+      let errors: Record<string, unknown> = {};
+
+      if (!comment) {
+        errors['comment'] = [
+          {
+            type: 'required',
+            value: comment,
+            message: 'Comment is required',
+          },
+        ];
+      }
+
+      if (!firstName) {
+        errors['firstName'] = [
+          {
+            type: 'required',
+            value: firstName,
+            message: 'First name is required',
+          },
+        ];
+      }
+
+      return Object.keys(errors).length === 0 ? undefined : errors;
+    };
 
     await render(<template>
       <ToucanForm @data={{data}} @validate={{formValidateCallback}} as |form|>
-        <form.Textarea @label="Comment" @name="comment" data-textarea />
+        <form.Textarea
+          @label="Comment"
+          @name="comment"
+          @rootTestSelector="data-textarea-wrapper"
```

Comment:
We use this to target individual error messages for the field! See selectors below